🎈🇩🇪 German language support

kellertuer opened the pull request:
In german there is a formal “Sie”-level of conversation and an informal/personal “Du”-Level. The Du is much more common these days even in some stores.
So for the software I also went for that. I used an AI (Claude to be precise) for the rough translation, but no worries, German is my mother tongue, so that was just to reduce a bit of typing. I then carenfully checked all strings again that they chose the right tone/form.
If one wants to be super exact the is a `de/de` (German/German) form, there might be subtle differences in de/at (Austrian) and de/ch (Swiss) German, but they should also just be subtle.

## A few comments and questions

* I kept order and line numbers the same for easier maintenance
* i dadpted the addition for languages from the Polish PR - should we order languages alphabetically in those places? Makes them easier to find when we have a few more.
* One thing I did not translate is “Notebook”. Formally that would be “Notizbuch” but for Pluto or  Jupyter Notebooks I have never heard anyone say Notizbuch; everyone says Notebook.
* The same for “Code” where also the AI kept that. The original word would be “Quelltext” but that is a bit like the 80s called...
* For some terminology like Stack-Trace there is no reasonable tranmslation and everyone would use Stack-Trace or souund from the 50s.
* both `t_stack_frame_this_cell` and `t_stack_frame_other_cell` are “This cell” in the original, is that on purpose? I would have called the second differently? 
* German has a few different capitalization rules in full sentences, I tried to adapt that carefully
* Though “⚠️ Danger” means “Gefahr” in the context of loading unsafe notebooks and such, I went for “⚠️ Achtung!”
* I feel AI and LLM are a bit mixed, but I kept their original usages.
* I am a bit unsure about `instantiate1` to 3, what do they mean? (lines 311-313)
* 314/315 run contains the verb evaluate evaluate contains Running - is that on purpose or to confuse the German (me) ;) ?
* For the lines 318-322 I am unsure wehther that reports ongoing actions (e.g. displayed _while_ connecting to backend) or when that is finished (_after having connected_) - in german those would translate differently

Finally I was a bit unsure how to test this - locally in pkg dev mode?

kellertuer commented on frontend/common/lang.js:
```diff
@@ -3,7 +3,7 @@ import LanguageDetector from "../imports/i18next-browser-languagedetector.js"
 import _ from "../imports/lodash.js"
 
-import { english, nederlands, ellinika } from "./lang_imports.js"
+import { english, nederlands, ellinika, deutsch } from "./lang_imports.js"
```
We could order this alphabetically?
```suggestion
import { deutsch, ellinika, english, nederlands } from "./lang_imports.js"
```

fonsp replied:
yes! you can already do that in your PR

kellertuer commented on frontend/common/lang.js:
```diff
@@ -22,8 +22,11 @@ i18next.use(LanguageDetector).init({
         },
+        de: {
+            translation: without_empty_keys(deutsch),
+        },
```
Also this whole block might be nicer ordered alphabetically, especially when we later have more languages.

kellertuer commented on frontend/lang/deutsch.json:
I went for `deutsch` here (that is German in German), since I was not totally sure which language the file name should be. The Greek indicated to maybe do it this way.

fonsp replied:
perfect!

fonsp commented:
Hey @kellertuer 

Thanks soooo much for your contribution!! This is super exciting 😍 I believe we have a lot of German-language users who will benefit a lot! This also goes great together with the German localization of PlutoTeachingTools.jl :)

For your comments – I feel like you 100% understood my intention with localization, and I agree with all your points :) The questions are really interesting, I had some similar choices with the Dutch localization and I made similar choices (like the informal "je", leaving tech words like "Notebook" and "Stack trace" in English, etc). 

# Some questions

> I used an AI (Claude to be precise) for the rough translation, but no worries, German is my mother tongue, so that was just to reduce a bit of typing. I then carenfully checked all strings again that they chose the right tone/form. 

Perfect! I forgot to write about using AI in the documentation, but I'm happy with it as long as all translations are checked by a human :)

> If one wants to be super exact the is a `de/de` (German/German) form, there might be subtle differences in de/at (Austrian) and de/ch (Swiss) German, but they should also just be subtle.

Good you mention it! I'd like to add nice support for local language variants at some point :) 

> * I kept order and line numbers the same for easier maintenance

Did you use the frontend/lang/update_languages.jl script also?

> * i adapted the addition for languages from the Polish PR - should we order languages alphabetically in those places? Makes them easier to find when we have a few more.

Sure!

> * both `t_stack_frame_this_cell` and `t_stack_frame_other_cell` are “This cell” in the original, is that on purpose? I would have called the second differently?

Oops! Well found :) I fixed it in TODOOOOOOOOO

> * I am a bit unsure about `instantiate1` to 3, what do they mean? (lines 311-313)

This is part of the "Status" tab in the bottom right.

I'm not sure! I think they all mean the same ("`Pkg.instantiate()` is running"). You can search for `:instantiate1` etc in our source code to be more detailed.

> * 314/315 run contains the verb evaluate evaluate contains Running - is that on purpose or to confuse the German (me) ;) ?

This is part of the "Status" tab in the bottom right.

😅 I looked in the source code (search for `:evaluate` and `:run`). "run" is the category that does a reactive evaluation of cells. It's structured like:

```
- run
  - resolve_topology (Pluto is analyzing code and reactive dependencies)
  - evaluate (Cell code is evaluated in the notebook process)
    - 1 (first cell)
    - 2 (second cell)
    - 3 etc
```

> * For the lines 318-322 I am unsure wehther that reports ongoing actions (e.g. displayed _while_ connecting to backend) or when that is finished (_after having connected_) - in german those would translate differently

This is part of the "Status" tab in the bottom right. So it describes a "subtask". It will show with the same text, first as "planned" (grey), then "running" (orange) and "done" (green). Do you need to write different text based on the status of the text? We could do that!

# How to test
> Finally I was a bit unsure how to test this - locally in pkg dev mode?

Exactly! You can follow the instructions here – https://github.com/fonsp/Pluto.jl/tree/main/frontend/lang#how-to-contribute

And you should be able to select German in the language selector.

fonsp commented:
It also looks like some of the text is flowing out of the box, e.g. the export menu:

<img width="895" height="175" alt="image" src="https://github.com/user-attachments/assets/12e9f3ae-5253-4cb7-9302-f62611674a60" />

This was to be expected! I designed this UI with hard-coded sizes for English text. 

I will see what to do with this tomorrow :) I will either make the boxes bigger, or I might modify the German languages files a bit and ask for your review

kellertuer commented:
> For your comments – I feel like you 100% understood my intention with localization, and I agree with all your points :) The questions are really interesting, I had some similar choices with the Dutch localization and I made similar choices (like the informal "je", leaving tech words like "Notebook" and "Stack trace" in English, etc).

Nice. In German stores the formal tone is still used every now and then but for example IKEA consistently does not do that. In Software I feel the move to informal is also relatively common

> Perfect! I forgot to write about using AI in the documentation, but I'm happy with it as long as all translations are checked by a human :)

Yes, I very carefully revised all translations. It was really meant to be a reduced typing issue.

> > If one wants to be super exact the is a `de/de` (German/German) form, there might be subtle differences in de/at (Austrian) and de/ch (Swiss) German, but they should also just be subtle.
> 
> Good you mention it! I'd like to add nice support for local language variants at some point :)

Nice. I do not know how to map that in a choice menu, but maybe also something to have in mind once that comes along

> > * I kept order and line numbers the same for easier maintenance
> 
> Did you use the frontend/lang/update_languages.jl script also?

Oh, sorry, no. My approach was to copy the English file and take it from there. Should that still be run or is it too late for that now? I ran it, and it would mainly delete a few empty lines in the file. For now I undid those changes again to keep the file aligned in lines with the English one.

> 
> > * both `t_stack_frame_this_cell` and `t_stack_frame_other_cell` are “This cell” in the original, is that on purpose? I would have called the second differently?
> 
> Oops! Well found :) I fixed it in [5852649](https://github.com/fonsp/Pluto.jl/commit/58526496aa59f5e28540145fb4fa996e515095fb)

Fixed it in the German version as well then.

> 
> > * I am a bit unsure about `instantiate1` to 3, what do they mean? (lines 311-313)
> 
> I'm not sure! I think they all mean the same ("`Pkg.instantiate()` is running"). You can search for `:instantiate1` etc in our source code to be more detailed.

Ok, I am not sure what they mean then and I will keep them as is. They might sound a bit strange if one is meant to be a plural or so. But in code it seems they all might mean the same.

> 
> > * 314/315 run contains the verb evaluate evaluate contains Running - is that on purpose or to confuse the German (me) ;) ?
> 
> This is part of the "Status" tab in the bottom right.
> 
> 😅 I looked in the source code (search for `:evaluate` and `:run`). "run" is the category that does a reactive evaluation of cells. It's structured like:
> 
> ```
> - run
>   - resolve_topology (Pluto is analyzing code and reactive dependencies)
>   - evaluate (Cell code is evaluated in the notebook process)
>     - 1 (first cell)
>     - 2 (second cell)
>     - 3 etc
> ```

This sadly does not help me much seeing what I should put in the translation

> 
>  > * For the lines 318-322 I am unsure wehther that reports ongoing actions (e.g. displayed _while_ connecting to backend) or when that is finished (_after having connected_) - in german those would translate differently
> 
> This is part of the "Status" tab in the bottom right. So it describes a "subtask". It will show with the same text, first as "planned" (grey), then "running" (orange) and "done" (green). Do you need to write different text based on the status of the text? We could do that!

I am not sure. Just from the English text I do not understand what a good translation would be (too less context) from your explanation I am still as unsure as before, sorry. But here is something in there in German that should maybe fit.

kellertuer commented:
> It also looks like some of the text is flowing out of the box [...]
>
> I will see what to do with this tomorrow :) I will either make the boxes bigger, or I might modify the German languages files a bit and ask for your review

Yeah one thing germans can do well is long words. But sure we can check what we can maybe shorten a bit as well.

kellertuer commented:
Thanks for the explanations. Then the current terms should be good.
